Audit logging beta documentation (#5834)
mattermost/docs: source/comply/audit-log.rst

Audit logging (beta)
====================

.. include:: ../_static/badges/ent-selfhosted.rst
  :start-after: :nosearch:

*Available in legacy Mattermost Enterprise Edition E20*

.. warning::

  - The audit logging beta is a **breaking change** from previous releases.
  - The format and content of an audit log record has changed to become standardized for all events.
  - Existing tools which ingest or parse audit log records may need to be modified.

Audit logging provides System Admins, including Security, IT/SRE, Compliance, and HR/PeopleOps teams, with a method of recording activities and events performed within a Mattermost workspace, such as access to the REST API or mmctl. Mattermost audit logging offers you control over where the audit logs are generated and stored, and applies a standard :doc:`JSON schema </comply/embedded-json-audit-log-schema>` to log output.

.. note::

  - Logs are recorded asynchronously to reduce latency to the caller, and are stored separately from general logging.

>   - During short spans of inability to write to targets, the audit records buffer in memory with a configurable maximum size cap. Based on typical audit record volumes, it could take many minutes to fill the buffer. After that, the records are dropped, and the record drop event is logged.

Reviewer: > audit records buffer in memory with a configurable maximum size cap

Should it be mentioned how to configure that?

Reviewer: Also, the `max size cap` is actually a maximum number of audit records buffered.  "Size" suggests memory size which is not the case.

Author: Yes. This is part of a follow-on effort to revisit logging documentation in general in a future release.

Author: Thanks! I've updated ``max size cap`` to ``max record size``. When Docs revisits logging documentation in general in a future release, we'll look to clarify in the docs how this cap is configured.

>   - During short spans of inability to write to targets, the audit records buffer in memory with a configurable maximum size cap. Based on typical audit record volumes, it could take many minutes to fill the buffer. After that, the records are dropped, and the record drop event is logged.

Author: ```suggestion
  - During short spans of inability to write to targets, the audit records buffer in memory with a configurable maximum record cap. Based on typical audit record volumes, it could take many minutes to fill the buffer. After that, the records are dropped, and the record drop event is logged.
```

| cert     | string   | Path to a cert file (.pem) to be used when establishing a TLS connection to the server.                                         |
+----------+----------+---------------------------------------------------------------------------------------------------------------------------------+
| insecure | bool     | Mattermost Boards accepts any certificate presented by the server, and any host name in that certificate. Default is ``false``. |
|          |          | **Note**: Should only be used in testing environments, and shouldn’t be used in production environments.                        |
+----------+----------+---------------------------------------------------------------------------------------------------------------------------------+
| tag      | string   | Syslog tag field.                                                                                                               |
+----------+----------+---------------------------------------------------------------------------------------------------------------------------------+

Format configuration options
----------------------------

> Plain log format configuration options

Reviewer: We don't actually specify what the available format types are: `plain`, `json`, `gelf`    A few do appear in the samples.

Author: Correct. This isn't yet comprehensive logging documentation. This PR focuses on audit logging only. Docs has a card on the backlog to revisit logging documentation in general in a future release.

*Available in legacy Mattermost Enterprise Edition E20*

.. warning::

  - The audit logging beta is a **breaking change** from previous releases.
  - The format and content of an audit log record has changed to become standardized for all events.
  - Existing tools which ingest or parse audit log records may need to be modified.

Audit logging provides System Admins, including Security, IT/SRE, Compliance, and HR/PeopleOps teams, with a method of recording activities and events performed within a Mattermost workspace, such as access to the REST API or mmctl. Mattermost audit logging offers you control over where the audit logs are generated and stored, and applies a standard :doc:`JSON schema </comply/embedded-json-audit-log-schema>` to log output.

.. note::

  - Logs are recorded asynchronously to reduce latency to the caller, and are stored separately from general logging.
  - During short spans of inability to write to targets, the audit records buffer in memory with a configurable maximum size cap. Based on typical audit record volumes, it could take many minutes to fill the buffer. After that, the records are dropped, and the record drop event is logged.

>   - When using syslog remotely, we recommend writing to a local file so no records are lost. Records won’t be taken from the local file and sent back to syslog when syslog becomes available again.

Reviewer: ```suggestion
  - When using syslog remotely, we recommend writing to a local file so no records are lost. Records will be taken from the local file and sent to the remote syslog when syslog becomes available again.
```

Reviewer: While I think this was the intent, I'm not sure where this functionality comes from.  Spooling to disk file and sending to remote server is not part of our implementation.

Reviewer: > When using syslog remotely, we recommend writing to a local file so no records are lost. Records will be taken from the local file and sent to the remote syslog when syslog becomes available again.

@cwarnermm  Are we sure this section is correct?  I don't know what is providing this support for spooling to disk file and having it sent later.  Some third party software would be needed for that.

Author: This is pre-existing content that was brought forward. Given your input, I propose we remove this line.

Configure audit logging
-----------------------

Configuring Mattermost to enable audit logging requires editing the ``config.json`` file directly. Audit logging can’t be managed using the System Console.

> In the ``config.json`` file, go to the ``ExperimentalAuditSettings`` section. Within the ``AdvancedLoggingConfig`` setting, you can specify an absolute or relative filespec to another configuration file, a database DSN, or a JSON string. The process of configuring audit logging includes specifying destination targets, event names to include, and the verbosity of the audit log output.

Reviewer: User can't set DSN themselves - that is only used when the whole config.json comes from a database.  I think it is enough to document file or escaped JSON string.  Otherwise users my think it can fetch config from any arbitrary database.

> In the ``config.json`` file, go to the ``ExperimentalAuditSettings`` section. Within the ``AdvancedLoggingConfig`` setting, you can specify an absolute or relative filespec to another configuration file, a database DSN, or a JSON string. The process of configuring audit logging includes specifying destination targets, event names to include, and the verbosity of the audit log output.

Author: ```suggestion
In the ``config.json`` file, go to the ``ExperimentalAuditSettings`` section. Within the ``AdvancedLoggingConfig`` setting, you can specify an absolute or relative filespec to another configuration file, or a JSON string. The process of configuring audit logging includes specifying destination targets, event names to include, and the verbosity of the audit log output.
```

        "type": "console",
        "format": "plain",
        "format_options": {
            "delim": " | "
        },
        "levels": [
          { "id": 100, "name": "audit-api" },
          { "id": 101, "name": "audit-content" },
          { "id": 102, "name": "audit-permissions" },
          { "id": 103, "name": "audit-cli" }
        ],
        "options": {
          "out": "stdout"
        },
        "maxqueuesize": 1000
      },
      "sample-file": {
        "type": "file",
        "format": "json",
        "levels": [
          { "id": 100, "name": "audit-api" },
          { "id": 101, "name": "audit-content" },
          { "id": 102, "name": "audit-permissions" },
          { "id": 103, "name": "audit-cli" }
        ],
        "options": {
          "compress": true,
          "filename": "audit.log",
          "max_age": 1,
          "max_backups": 10,
          "max_size": 500
        },
        "maxqueuesize": 1000
      }
    }

Examples of values for the ``AdvancedLoggingConfig`` setting are:

1. Filespec to another configuration file; this file will contain a JSON object

  ``"AdvancedLoggingConfig": "/path/to/audit_log_config.json"``

2. Database DSN

>   ``"AdvancedLoggingConfig": "postgres://user@host:5432/database"``

Reviewer: This part is not needed as per above comment.

3. JSON string

  ``"AdvancedLoggingConfig": "{\"sample-console\":{\"type\":\"console\",\"format\":\"plain\",\"format_options\":{\"delim\":\" | \"},\"levels\":[{\"id\":100,\"name\":\"audit-api\"},{\"id\":101,\"name\":\"audit-content\"},{\"id\":102,\"name\":\"audit-permissions\"},{\"id\":103,\"name\":\"audit-cli\"}],\"options\":{\"out\":\"stdout\"},\"maxqueuesize\":1000},\"sample-file\":{\"type\":\"file\",\"format\":\"json\",\"levels\":[{\"id\":100,\"name\":\"audit-api\"},{\"id\":101,\"name\":\"audit-content\"},{\"id\":102,\"name\":\"audit-permissions\"},{\"id\":103,\"name\":\"audit-cli\"}],\"options\":{\"compress\":true,\"filename\":\"audit.log\",\"max_age\":1,\"max_backups\":10,\"max_size\":500},\"maxqueuesize\":1000}}"``

.. note::

>   When using a JSON string as the value of ``AdvancedLoggingConfig``, ensure you escape double quotes (``"``) in the string using a backslash (``\``).

Author: ```suggestion
  When using a JSON string as the value of ``AdvancedLoggingConfig``, ensure you escape double quotes (``"``) in the string using a backslash (``\``). You can also use a free online tool, such as `Free Online JSON Escape <https://www.freeformatter.com/json-escape.html>`__ to format the value correctly.
```

|            |          | - Black: ``30``                                              |
|            |          | - Red: ``31``                                                |
|            |          | - Green: ``32``                                              |
|            |          | - Yellow: ``33``                                             |
|            |          | - Blue: ``34``                                               |
|            |          | - Magenta: ``35``                                            |
|            |          | - Cyan: ``36``                                               |
|            |          | - White: ``37``                                              |
+------------+----------+--------------------------------------------------------------+

Audit log levels
~~~~~~~~~~~~~~~~

The following audit log levels are available:

+--------+-----------------------+--------------------+
| **ID** | **Name**              | **Description**    |
+--------+-----------------------+--------------------+
| 100    | ``audit-api``         | API events         |
+--------+-----------------------+--------------------+

> | 101    | ``audit-content``     | Content changes    |

Reviewer: Can we add a note regarding `audit-content` level denoting that it can generate a *lot* more records than the others?

+--------+-----------------------+--------------------+
| 102    | ``audit-permissions`` | Permission changes |
+--------+-----------------------+--------------------+
| 103    | ``audit-cli``         | CLI operations     |
+--------+-----------------------+--------------------+

Multiple file and target support
--------------------------------

System Admins can define multiple log targets to:

- Mirror log output to files and log aggregators for redundancy.

> - Log certain entries to specific destinations. For example, all errors could be routed to a specific destination for review.

Reviewer: Sending errors to a specific destination applies to general logging, but for audits only the four levels above are used (i.e. no `error` level).  This can be re-worded to something like: 

```suggestion
- Log certain entries to specific destinations. For example, all `audit-content` records could be routed to a different destination than the other levels.
```

Admins can also temporarily disable log targets by setting its type to ``none``.

> Any combination of console, local file, syslog, and TCP socket targets can send log records to multiple targets. Log targets have been chosen based on support for the vast majority of log aggregators and other log analysis tools, without needing additional software installed.

Reviewer: > Any combination of console, local file, syslog, and TCP socket targets can send log records to multiple targets.

I feel this phrasing is a bit awkward.  :man_shrugging:

Author: Thanks! Simplified this sentence.

| disables_stacktrace | bool     | Disables output of stack traces. Default is ``false``.                                  |
+---------------------+----------+-----------------------------------------------------------------------------------------+
| timestamp_format    | string   | Format for timestamps. Default is `RFC3339 <https://www.rfc-editor.org/rfc/rfc3339>`__. |
+---------------------+----------+-----------------------------------------------------------------------------------------+

GELF log format configuration options
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+----------+----------+----------------------------------------------------------+
| **Key**  | **Type** | **Description**                                          |
+----------+----------+----------------------------------------------------------+
| hostname | string   | Outputs a custom hostname in log records.                |
|          |          | If omitted, hostname is taken from the operating system. |
+----------+----------+----------------------------------------------------------+

Configure audit logging in Mattermost Boards
--------------------------------------------

> The `Boards configuration file <https://github.com/mattermost/focalboard/blob/main/config.json>`_ ``config.json`` is used to configure logging. There are two settings which control logging:

Reviewer: This is only for FocalBoard personal server and desktop edition.  Does not apply to Boards within the suite.

Reviewer: This whole Boards section doesn't apply to suite.  I suggest it be removed.

Author: @wuwinson - Looping you in on Doug's suggestion. Which route would you prefer the Audit Logging documentation take with regards to Boards? Specifying that the content applies _only) to the personal server and desktop edition, OR removing this section of the docs entirely?

Reviewer 2: Agreed with @wiggin77 - lets remove it so the docs don't confuse our users.
